shared/lib_battery_lifetime_calendar_cycle.cpp

```diff
@@ -42,9 +42,25 @@ void lifetime_cycle_t::initialize() {
     state->cycle->rainflow_Xlt = 0;
     state->cycle->rainflow_Ylt = 0;
     state->cycle->rainflow_peaks.clear();
+    init_cycle_counts();
     resetDailyCycles();
 }
 
+void lifetime_cycle_t::init_cycle_counts() {
+    std::vector<double> DOD_levels;
+    for (size_t i = 0; i < params->cal_cyc->cycling_matrix.nrows(); i++) {
+        double dod = params->cal_cyc->cycling_matrix.at(i, calendar_cycle_params::DOD);
+        if (std::find(DOD_levels.begin(), DOD_levels.end(), dod) == DOD_levels.end()) {
+            DOD_levels.push_back(dod);
+        }
+    }
+    std::sort(DOD_levels.begin(), DOD_levels.end());
+    state->cycle->cycle_counts.resize_fill(DOD_levels.size(), 2, 0.0);
+    for (size_t i = 0; i < DOD_levels.size(); i++) {
+        state->cycle->cycle_counts.set_value(DOD_levels[i], i, cycle_state::DOD);
+    }
```

Comment:
does it make sense to restrict the reporting of the cycle_count matrices' DOD levels to what is provided in the cycling_matrix? This isn't how the cycle model actually works because it interpolates between DODs. I may prefer a fixed number of equal-width bins for reporting (like 10: 0-9, 10-19, ... 90-99).

Reply:
This is used by the model in lines 183 through 192 to determine how many cycles to apply in the bilinear interpolation. I was assuming that the user would specify any non-linear dynamics on different DOD vs cycle curves, and that maintaining the bins specified on chart would allow us to hit roughly the right point on each line.

I'm open to reporting smaller bins in a second output variable, but I think the correlation with the chart is worth maintaining if we're using cycle_counts in the degradation computation.